### PR TITLE
MCR-2517 Expand the service section of objects to store classifications and messages

### PR DESCRIPTION
**Pull request by toKrause**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2517).

### CODE REVIEW

**Review by rsteph-de — COMMENTED** [review]
It could be better to extend MCRMetaLangText

**Review by rsteph-de — COMMENTED** [review]
We need to extend the datamodel plugin
that the new data types can also be used in the metadata section:

https://github.com/MyCoRe-Org/datamodel-plugin/blob/main/src/main/resources/datamodel2schema.xsl

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 1]
```diff
@@ -840,4 +927,371 @@ public final int getFlagIndex(String value) {
...
+    }
+
+    /**
+     * This method returns all classification values of the specified type.
+     *
+     * @param classId
+     *              a classId as string.
+     * @return a list of classification values
+     */
+    public final List<MCRCategoryID> getClassificationsAsMCRCategoryID(String classId) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: This should be changed to 
List<MCRCategoryID> getClassifications(String type)

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 2]
```diff
@@ -840,4 +927,371 @@ public final int getFlagIndex(String value) {
...
+
+    /**
+     * This method get a single classification from the classification list as a string.
+     *
+     * @exception IndexOutOfBoundsException
+     *                throw this exception, if the index is invalid
+     * @return a classification string
+     */
+    public final String getClassification(int index) throws IndexOutOfBoundsException {
+        if (index < 0 || index >= classifications.size()) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java:1196]: always return MCRCategoryID for classification (categid as String is 'incomplete')

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 3]
```diff
@@ -840,4 +927,371 @@ public final int getFlagIndex(String value) {
...
+    }
+
+    /**
+     * This method adds a classification to the classifications list.
+     *
+     * @param classification
+     *              the new classification value as string
+     */
+    public final void addClassification(MCRCategoryID classification) {
+        if (classification == null) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: do the actual "adding" in this method and call this method in addClassfication(String, String)

**Review by rsteph-de — CHANGES_REQUESTED** [review]
some initial ideas ...

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java (hunk 1 above)
Comment (toKrause) [reply mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: Why `type`? Classifiactions have a `classId` and a `categId` (or a `rootId` and an `id`, depending on the context). Why introduce yet another nomenclature?

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java (hunk 2 above)
Comment (toKrause) [reply mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java:1196]: I've actually raised this as a talking point (access to classifiactions vs. access to other multi-property-entities in the service section, like flags) in my mail to the developer mailing list this Monday. I'd prefer some kind of consistency. I hope we find time to discuss this in the next web conference.

File: mycore-base/src/main/java/org/mycore/datamodel/common/MCRLinkTableManager.java [hunk 4]
```diff
@@ -475,6 +475,9 @@ public void create(MCRObject obj) {
         if (state != null) {
             categories.add(state);
         }
+        for(MCRCategoryID classification : obj.getService().getClassifications()) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/common/MCRLinkTableManager.java]: shorter: categories.addAll(object.getService().getClassifications())

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 5]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+    public final int getMessagesSize() {
+        return messages.size();
+    }
+
+    /**
+     * Returns the messages as list.
+     *
+     * @return messages as list
+     */
+    protected final List<MCRMetaDateLangText> getMessagesAsList() {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: The method name is inconsistent:
better:  getMessagesAsMCRMetaLangText()

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 6]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+    }
+
+    /**
+     * This method returns all message values of the specified type.
+     *
+     * @param type
+     *              a type as string.
+     * @return a list of message values
+     */
+    protected final List<MCRMetaDateLangText> getMessagesAsMCRMetaLangText(String type) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: better: gerMessagesAsMCRMetaLangTextForType(...)

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 7]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+
+    /**
+     * This method returns all messages values of the specified type.
+     *
+     * @param type
+     *              a type as string.
+     * @return a list of message values
+     */
+    public final List<String> getMessages(String type) {
+        return getMessagesAsMCRMetaLangText(type).stream()
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: better: getMessagesForType(...)

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 8]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+    }
+
+    /**
+     * This method removes all messages of the specified type from
+     * the message list.
+     *
+     * @param type
+     *            a type as string
+     */
+    public final void removeMessages(String type) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java:1128]: better:removeMessagesForType(...)

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 9]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+    public final int getClassificationsSize() {
+        return this.classifications.size();
+    }
+
+    /**
+     * This method returns all classification values.
+     *
+     * @return classifications as list
+     */
+    protected final List<MCRMetaClassification> getClassificationsAsList() {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: better: getClassificationsAsMCRMetaClassification(...)

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 10]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+    }
+
+    /**
+     * This method returns all classification values of the specified type.
+     *
+     * @param type
+     *              a type as string.
+     * @return a list of classification values
+     */
+    protected final List<MCRMetaClassification> getClassificationsAsMCRMetaClassification(String type) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: better: ...ForType(...)

File: mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java [hunk 11]
```diff
@@ -840,4 +927,380 @@ public final int getFlagIndex(String value) {
...
+    }
+
+    /**
+     * This method returns all classification values of the specified type.
+     *
+     * @param type
+     *              a type as string.
+     * @return a list of classification values
+     */
+    public final List<MCRCategoryID> getClassifications(String type) {
```
Comment (rsteph-de) [inline mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRObjectService.java]: ...ForType(...)

**Review by rsteph-de — CHANGES_REQUESTED** [review]
good work, but I would suggest some changes to the method names
in MCRObjectService to be more consistent

**Comment by toKrause** [comment]
I fail to see large numbers of inconsistencies in the PR. The proposed changes would actually introduce a lot of inconsistencies in the API design. I modeled the new method signatures after the existing method signatures for the flags.

---

We already have:
`protected final List<MCRMetaLangText> getFlagsAsList()`

Hence:
`protected final List<MCRMetaDateLangText> getMessagesAsList()`
`protected final List<MCRMetaClassification> getClassificationsAsList()`

---

We already have:
`protected final ArrayList<MCRMetaLangText> getFlagsAsMCRMetaLangText(String type)`

Hence:
`protected final List<MCRMetaDateLangText> getMessagesAsMCRMetaLangText(String type)`
`protected final List<MCRMetaClassification> getClassificationsAsMCRMetaClassification(String type)`

---

We already have:
`public final ArrayList<String> getFlags(String type)`

Hence:
`public final List<String> getMessages(String type)`
`public final List<MCRCategoryID> getClassifications(String type)`

---

We already have:
`public final void removeFlags(String type)`

Hence:
`public final void removeMessages(String type)`
`public final void removeClassifications(String type)`

---

There are only two inconsistencies; which I will fix

- I return `List` instead of `ArrayList` (which is arguably better, because it is less specific coupling)
- `getMessagesAsMCRMetaLangText` should be `getMessagesAsMCRMetaDateLangText`

**Comment by toKrause** [comment]
test failed because "Unable to find commits until some tag: Walk failure. Missing commit 8a2a3f74dbcd27c5bf94d4176914d3bd6b7a43e5". please restart

**Comment by rsteph-de** [comment]
The names of methods dealing with flags are little bit confusing:
current state:
public final String getFlags() //flags as String (concated by " ")    
List<String> getFlags(String type)
List<MCRMetaLangText> getFlagsAsList()
List<MCRMetaLangText> getFlagsAsMCRMetaLangText(String type) 

my wish:
String getFlagsAsString()  (or better replaced with the following)
List<String> getFlags()
List<String> getFlags(String type)
List<MCRMetaLangText> getFlagsAsMCRMetaLangText()
List<MCRMetaLangText> getFlagsAsMCRMetaLangText(String type) 

But this PR is not the place to clean it up. 
So I'll approve your changes after the build is successful.

**Review by rsteph-de — APPROVED** [review]
O.K.

**Pull request merged**